src/QsCompiler/CompilationManager/EditorCompletion.cs

```diff
@@ -0,0 +1,686 @@
+            if (relativeLine < 0 ||
+                relativeLine >= lines.Count() ||
+                relativeCharacter < 0 ||
+                relativeCharacter > lines.ElementAt(relativeLine).Length)
+            {
+                throw new ArgumentException("Position is outside the fragment range", nameof(position));
+            }
+            return lines.Take(relativeLine).Sum(line => line.Length) + relativeCharacter;
+        }
+
+        /// <summary>
+        /// Resolves the namespace alias and returns its full namespace name. If the alias couldn't be resolved, returns
+        /// the alias unchanged.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static string ResolveNamespaceAlias(
+            FileContentManager file, CompilationUnit compilation, Position position, string alias)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (compilation == null)
+                throw new ArgumentNullException(nameof(compilation));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+            if (alias == null)
+                throw new ArgumentNullException(nameof(alias));
+
+            var nsName = file.TryGetNamespaceAt(position);
+            if (nsName == null)
+                return alias;
+            return compilation.GlobalSymbols.TryResolveNamespaceAlias(
+                NonNullable<string>.New(alias), NonNullable<string>.New(nsName), file.FileName) ?? alias;
+        }
+
+        /// <summary>
+        /// Gets the token index and corresponding code fragment at, or the closest before, the given position in the
+        /// file. Returns the relationship between the found fragment and the fragment that would exist directly at the
+        /// given position, which may be different from <see cref="FragmentRelationship.Self"/> if the given position is
+        /// not part of any fragment.
+        /// <para/>
+        /// If <see cref="FragmentRelationship.Missing"/> is returned, then both the token index and code fragment out
+        /// parameters are set to null.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">Thrown when any argument is null.</exception>
+        private static FragmentRelationship GetFragmentAtOrBefore(
+            FileContentManager file,
+            Position position,
+            out CodeFragment.TokenIndex index,
+            out CodeFragment fragment)
+        {
+            if (file == null)
+                throw new ArgumentNullException(nameof(file));
+            if (position == null)
+                throw new ArgumentNullException(nameof(position));
+
+            var lineNumber = position.Line;
+            var tokens = file.GetTokenizedLine(lineNumber);
+
+            // If the current line is empty, find the last non-empty line before it.
```

Comment:
I there are a bunch of similar things implemented in the TextProcessor, so you may be able to use one of those if a suitable one already exists (I don't remember if a suitable one exists, but it seems like something that might).

Reply:
We weren't able to find an existing function for `GetTokenAtOrBefore`, but I was able to replace most of the logic in `GetDelimiterPosition` with `FragmentEnd` from TextProcessor.cs.